src/app/modules/form/admin-form/admin-form.controller.ts

```diff
@@ -1717,6 +1718,60 @@ export const _handleReorderFormField: RequestHandler<
       })
   )
 }
+/*
+ * Handler for PUT /forms/:formId/logic/:logicId
+ * @security session
+ *
+ * @returns 200 with success message when successfully updated
+ * @returns 403 when user does not have permissions to update logic
+ * @returns 404 when form cannot be found
+ * @returns 422 when user in session cannot be retrieved from the database
+ * @returns 500 when database error occurs
+ */
+export const handleUpdateLogic: RequestHandler<
+  ParamsDictionary,
+  unknown,
+  { updatedLogic: ILogicSchema }
+> = (req, res) => {
+  const { formId, logicId } = req.params
+  const { updatedLogic } = req.body
+  const sessionUserId = (req.session as Express.AuthedSession).user._id
+
+  // Step 1: Retrieve currently logged in user.
+  return (
+    UserService.getPopulatedUserById(sessionUserId)
+      .andThen((user) =>
+        // Step 2: Retrieve form with write permission check.
+        AuthService.getFormAfterPermissionChecks({
+          user,
+          formId,
+          level: PermissionLevel.Write,
+        }),
+      )
+
+      // Step 3: Update form logic
+      .andThen((retrievedForm) =>
+        AdminFormService.updateFormLogic(retrievedForm, logicId, updatedLogic),
+      )
+      .map(() => res.sendStatus(StatusCodes.OK))
```

Comment:
we should be returning the saved updated logic field so the client always gets a copy which is consistent with the current state of the db

Reply:
edited